# Tugas

## Tugas 1

Buatlah model klasfikasi Multinomial Naive Bayes dengan ketentuan,

1. Menggunakan data `spam.csv`
2. Fitur `CountVectorizer` dengan mengaktifkan **stop_words**
3. Evaluasi hasilnya

## Tugas 2

Buatlah model klasfikasi Multinomial Naive Bayes dengan ketentuan,

1. Menggunakan data `spam.csv`
2. Fitur `TF-IDF` dengan mengaktifkan **stop_words**
3. Evaluasi hasilnya dan bandingkan dengan hasil tugas 1.
4. Berikan kesimpulan fitur mana yang terbaik pada kasus data `spam.csv`

# Tugas 1

In [2]:
import numpy as np
import pandas as pd
df = pd.read_csv('spam.csv', encoding='latin-1') 
df.head()

v1                                                 v2 Unnamed: 2  \
0   ham  Go until jurong point, crazy.. Available only ...        NaN   
1   ham                      Ok lar... Joking wif u oni...        NaN   
2  spam  Free entry in 2 a wkly comp to win FA Cup fina...        NaN   
3   ham  U dun say so early hor... U c already then say...        NaN   
4   ham  Nah I don't think he goes to usf, he lives aro...        NaN   

  Unnamed: 3 Unnamed: 4  
0        NaN        NaN  
1        NaN        NaN  
2        NaN        NaN  
3        NaN        NaN  
4        NaN        NaN

In [3]:
df = df.drop(df.iloc[:,2:], axis=1)
df.head()

v1                                                 v2
0   ham  Go until jurong point, crazy.. Available only ...
1   ham                      Ok lar... Joking wif u oni...
2  spam  Free entry in 2 a wkly comp to win FA Cup fina...
3   ham  U dun say so early hor... U c already then say...
4   ham  Nah I don't think he goes to usf, he lives aro...

In [4]:
new_cols = {
    'v1': 'Kategori',
    'v2': 'Pesan'
}
df = df.rename(columns=new_cols)
df.head()

Kategori                                              Pesan
0      ham  Go until jurong point, crazy.. Available only ...
1      ham                      Ok lar... Joking wif u oni...
2     spam  Free entry in 2 a wkly comp to win FA Cup fina...
3      ham  U dun say so early hor... U c already then say...
4      ham  Nah I don't think he goes to usf, he lives aro...

In [6]:
# Cek Jumlah Data Per Kelas
print(df['Kategori'].value_counts())
print('\n')
# Cek Kelengkapan Data
print(df.info())
print('\n')
# Cek Statistik Deskriptif
print(df.describe())

ham     4825
spam     747
Name: Kategori, dtype: int64


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5572 entries, 0 to 5571
Data columns (total 2 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   Kategori  5572 non-null   object
 1   Pesan     5572 non-null   object
dtypes: object(2)
memory usage: 87.2+ KB
None


       Kategori                   Pesan
count      5572                    5572
unique        2                    5169
top         ham  Sorry, I'll call later
freq       4825                      30


In [7]:
# Data untuk label
new_labels = {
    'spam': 1,
    'ham': 0
}
# Encode label
df['Kategori'] = df['Kategori'].map(new_labels)
# Cek data
df.head()

Kategori                                              Pesan
0         0  Go until jurong point, crazy.. Available only ...
1         0                      Ok lar... Joking wif u oni...
2         1  Free entry in 2 a wkly comp to win FA Cup fina...
3         0  U dun say so early hor... U c already then say...
4         0  Nah I don't think he goes to usf, he lives aro...

In [23]:
X = df['Pesan'].values
y = df['Kategori'].values
print(X)
print(y)

['Go until jurong point, crazy.. Available only in bugis n great world la e buffet... Cine there got amore wat...'
 'Ok lar... Joking wif u oni...'
 "Free entry in 2 a wkly comp to win FA Cup final tkts 21st May 2005. Text FA to 87121 to receive entry question(std txt rate)T&C's apply 08452810075over18's"
 ... 'Pity, * was in mood for that. So...any other suggestions?'
 "The guy did some bitching but I acted like i'd be interested in buying something else next week and he gave it to us for free"
 'Rofl. Its true to its name']
[0 0 1 ... 0 0 0]


In [27]:
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import CountVectorizer
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=50)
bow = CountVectorizer(stop_words='english')
X_train = bow.fit_transform(X_train)
X_test = bow.transform(X_test)
# X_train = bow.get_feature_names()
# X_test = bow.get_feature_names()
# print(len(bow.get_feature_names()))

7466


In [28]:
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import accuracy_score

# Inisiasi MultinomialNB
mnb = MultinomialNB()

# Fit model
mnb.fit(X_train, y_train)

# Prediksi dengan data training
y_pred_train = mnb.predict(X_train)

# Evaluasi akurasi data training
acc_train = accuracy_score(y_train, y_pred_train)

# Prediksi dengan data training
y_pred_test = mnb.predict(X_test)

# Evaluasi akurasi data training
acc_test = accuracy_score(y_test, y_pred_test)

# Print hasil evaluasi
print(f'Hasil akurasi data train: {acc_train}')
print(f'Hasil akurasi data test: {acc_test}')

Hasil akurasi data train: 0.9946152120260264
Hasil akurasi data test: 0.9829596412556054


# Tugas 2

In [34]:
from sklearn.feature_extraction.text import TfidfVectorizer
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=50)
bow2 = TfidfVectorizer(stop_words='english')
X_train = bow2.fit_transform(X_train)
X_test = bow2.transform(X_test)
X_train = bow.get_feature_names()
X_test = bow.get_feature_names()
# print(X_train)
# print(y_test)
print(len(bow.get_feature_names()))

7466


In [31]:
# Inisiasi MultinomialNB
mnb = MultinomialNB()

# Fit model
mnb.fit(X_train, y_train)

# Prediksi dengan data training
y_pred_train = mnb.predict(X_train)

# Evaluasi akurasi data training
acc_train = accuracy_score(y_train, y_pred_train)

# Prediksi dengan data training
y_pred_test = mnb.predict(X_test)

# Evaluasi akurasi data training
acc_test = accuracy_score(y_test, y_pred_test)

# Print hasil evaluasi
print(f'Hasil akurasi data train: {acc_train}')
print(f'Hasil akurasi data test: {acc_test}')

Hasil akurasi data train: 0.9842943684092439
Hasil akurasi data test: 0.9605381165919282


### Count Vectorizer tanpa stop_words
Hasil akurasi data train: 0.9946152120260264

Hasil akurasi data test: 0.9775784753363229
### Count Vectorizer with stop_words
Hasil akurasi data train: 0.9946152120260264

Hasil akurasi data test: 0.9829596412556054
### TF-IDF
Hasil akurasi data train: 0.9842943684092439

Hasil akurasi data test: 0.9605381165919282

## Kesimpulan
Count Vectorizer tanpa stop_words memiliki data test yang *lebih kecil* daripada Count Vectorizer with stop_words, karena data test untuk Count Vectorizer tanpa stop_words(7727) *lebih banyak* dari pada data test untuk Count Vectorizer with stop_words(7466) yang mana berpengaruh dengan hasil testing data spam. Jadi lebih baik menggunakan stop words untuk untuk hasil yang lebih akurat.

Untuk Count Vectorizer memiliki hasil train dan testing yang lebih akurat dari pada TF-IDF. Hal ini kemungkinan dikarenakan TF-IDF lebih rinci dalam training dan testingnya sehingga menyebabkan probabilitas yang lebih rendah dari pada count vectorizer. Jadi untuk probobilitas lebih baik atau lebih tinggi Count Vectorizer dengan stop_word.